cpp/src/arrow/compute/kernels/scalar_arithmetic.cc

```diff
@@ -516,6 +564,190 @@ std::shared_ptr<ScalarFunction> MakeUnarySignedArithmeticFunctionNotNull(
   return func;
 }
 
+using MinMaxState = OptionsWrapper<ElementWiseAggregateOptions>;
+
+// Implement a variadic scalar min/max kernel.
+template <typename OutType, typename Op>
+struct ScalarMinMax {
+  using OutValue = typename GetOutputType<OutType>::T;
+
+  static void ExecScalar(const ExecBatch& batch,
+                         const ElementWiseAggregateOptions& options, Scalar* out) {
+    // All arguments are scalar
+    OutValue value{};
+    bool valid = false;
+    for (const auto& arg : batch.values) {
+      // Ignore non-scalar arguments so we can use it in the mixed-scalar-and-array case
+      if (!arg.is_scalar()) continue;
+      const auto& scalar = *arg.scalar();
+      if (!scalar.is_valid) {
+        if (options.skip_nulls) continue;
+        out->is_valid = false;
+        return;
+      }
+      if (!valid) {
+        value = UnboxScalar<OutType>::Unbox(scalar);
+        valid = true;
+      } else {
+        value = Op::Call(value, UnboxScalar<OutType>::Unbox(scalar));
+      }
+    }
+    out->is_valid = valid;
+    if (valid) {
+      BoxScalar<OutType>::Box(value, out);
+    }
+  }
+
+  static Status Exec(KernelContext* ctx, const ExecBatch& batch, Datum* out) {
+    const ElementWiseAggregateOptions& options = MinMaxState::Get(ctx);
+    const auto descrs = batch.GetDescriptors();
+    const size_t scalar_count =
+        static_cast<size_t>(std::count_if(batch.values.begin(), batch.values.end(),
+                                          [](const Datum& d) { return d.is_scalar(); }));
+    if (scalar_count == batch.values.size()) {
+      ExecScalar(batch, options, out->scalar().get());
+      return Status::OK();
+    }
+
+    ArrayData* output = out->mutable_array();
```

Comment:
Instead of trying to recycle buffers, we *could* change the kernel's flag and allow the executor to preallocate the output data buffer. Then we populate this with the anti extreme (for maximum/uint64_t we zero the buffer). Then we wouldn't need to use a bitmap to express "no values folded yet"; `get_extreme({anti_extreme, v...}) == get_extreme({v...})`. We'd also be able to compute the null bitmap separately in either case: `!skip_nulls` AND the bitmaps, `skip_nulls` OR the bitmaps.

In the case where `scalar_count != 0`, we can ExecScalar and use the result instead of an anti_extreme

Reply:
This works quite well except in the case that we have a singular NaN array argument (`element_wise_min([NaN])`) - in which case we want to emit `NaN` not the antiextreme.

Reply:
Ah, maybe let's just have the 'antiextreme' of a float be NaN.

Reply:
The antiextremes of float are Inf, -Inf

Reply:
Inf would trip up if we had something like `element_wise_min([NaN, 0], [NaN, 1])` in which case we'd want to emit `NaN` and not `Inf` as the first element. Using `Inf` as the antiextreme means we'd emit `Inf` since the kernel takes any non-NaN value over any NaN.

Reply:
Huh, I was expecting `fmin(NaN, x) = NaN` but in fact it's `fmin(NaN, x) = x`. Nevermind